### PR TITLE
Check for SIGHUP

### PR DESCRIPTION
**Pull request by r00ster91**
If you have the `termination` feature enabled and close the terminal (its window) with the X button, `SIGHUP` is sent to the process but `ctrlc` doesn't handle it so no graceful shutdown is done in that case but I would expect the `termination` feature to be able to detect that because pressing the X button to terminate an application is definitely a common kind of termination. Also, I believe on Windows this already works because when the window is X-ed, `CTRL_CLOSE_EVENT` is sent which `ctrlc` seems to handle. This means on Windows, the handler is invoked on X but on Linux (and very likely macOS), it is not. So this fixes inconsistent behavior.

### CODE REVIEW

File: src/platform/unix/mod.rs [hunk 1]
```diff
@@ -105,11 +105,20 @@ pub unsafe fn init_os_handler() -> Result<(), Error> {
...
+            Ok(_) => {}
+            Err(e) => {
+                signal::sigaction(signal::Signal::SIGINT, &_old).unwrap();
+                return Err(close_pipe(e));
+            }
+        }
+        match signal::sigaction(signal::Signal::SIGHUP, &new_action) {
+            Ok(_) => {}
+            Err(e) => {
+                signal::sigaction(signal::Signal::SIGINT, &_old).unwrap();
```
Comment (Detegr) [inline src/platform/unix/mod.rs:119]: Old value for SIGTERM should be reverted in this cleanup function. Also, the variable `_old` should be named `old` now as it is being used. Maybe assign current `_old` to `sigint_old` and introduce `sigterm_old`.

Comment (r00ster91) [reply src/platform/unix/mod.rs]: I made a change. Is that maybe what you have in mind?

**Comment by r00ster91** [comment]
I took another look at https://en.wikipedia.org/wiki/Signal_(IPC)#POSIX_signals and found this other signal:
> **SIGTSTP**
    The SIGTSTP signal is sent to a process by its controlling **terminal** to request it to **stop** (**t**erminal **st**o**p**). It is commonly initiated by the user pressing `Ctrl`+`Z`. Unlike SIGSTOP, the process can register a signal handler for, or ignore, the signal.

This should probably be handled as well to ensure a graceful termination whenever possible.

Do you see anything else that might be worth checking for on the article?

File: src/platform/unix/mod.rs [hunk 2]
```diff
@@ -99,17 +99,27 @@ pub unsafe fn init_os_handler() -> Result<(), Error> {
...
     };
 
     #[cfg(feature = "termination")]
-    match signal::sigaction(signal::Signal::SIGTERM, &new_action) {
-        Ok(_) => {}
-        Err(e) => {
-            signal::sigaction(signal::Signal::SIGINT, &_old).unwrap();
-            return Err(close_pipe(e));
+    {
+        let sigterm_old = sigint_old;
```
Comment (Detegr) [inline src/platform/unix/mod.rs:109]: This is not quite what I meant.

File: src/platform/unix/mod.rs [hunk 3]
```diff
@@ -99,17 +99,27 @@ pub unsafe fn init_os_handler() -> Result<(), Error> {
         signal::SigSet::empty(),
     );
 
-    let _old = match signal::sigaction(signal::Signal::SIGINT, &new_action) {
+    let sigint_old = match signal::sigaction(signal::Signal::SIGINT, &new_action) {
```
Comment (Detegr) [inline src/platform/unix/mod.rs:102]: This is good :+1:

File: src/platform/unix/mod.rs [hunk 4]
```diff
@@ -99,17 +99,27 @@ pub unsafe fn init_os_handler() -> Result<(), Error> {
...
-            return Err(close_pipe(e));
+    {
+        let sigterm_old = sigint_old;
+        match signal::sigaction(signal::Signal::SIGTERM, &new_action) {
+            Ok(_) => {}
+            Err(e) => {
+                signal::sigaction(signal::Signal::SIGINT, &sigint_old).unwrap();
+                return Err(close_pipe(e));
+            }
+        }
```
Comment (Detegr) [inline src/platform/unix/mod.rs:116]: This should look like the one that handles SIGINT, so something like:
```
let sigterm_old = match signal::sigaction(signal::Signal::SIGTERM ...
    Ok(old) => {old}
    Err(e) => { ... }
 ```

File: src/platform/unix/mod.rs [hunk 5]
```diff
@@ -99,17 +99,27 @@ pub unsafe fn init_os_handler() -> Result<(), Error> {
...
+        match signal::sigaction(signal::Signal::SIGHUP, &new_action) {
+            Ok(_) => {}
+            Err(e) => {
+                signal::sigaction(signal::Signal::SIGINT, &sigterm_old).unwrap();
```
Comment (Detegr) [inline src/platform/unix/mod.rs:120]: Here the cleanup should be:
```
signal::sigaction(signal::Signal::SIGINT, &sigint_old).unwrap();
signal::sigaction(signal::Signal::SIGTERM, &sigterm_old).unwrap();
```

**Review by Detegr — CHANGES_REQUESTED** [review]

**Comment by Detegr** [comment]
> I took another look at https://en.wikipedia.org/wiki/Signal_(IPC)#POSIX_signals and found this other signal:
> 
> > **SIGTSTP**
> > The SIGTSTP signal is sent to a process by its controlling **terminal** to request it to **stop** (**t**erminal **st**o**p**). It is commonly initiated by the user pressing `Ctrl`+`Z`. Unlike SIGSTOP, the process can register a signal handler for, or ignore, the signal.
> 
> This should probably be handled as well to ensure a graceful termination whenever possible.
> 
> Do you see anything else that might be worth checking for on the article?
> 
> Maybe `SIGABRT` and `SIGQUIT`?

I don't thing `SIGSTOP` should be handled. It is not a termination, but just a temporary stop for a process that is going to be continued later. `SIGABRT` and `SIGQUIT` are program error signals, and I don't think it's a good idea to add a handler for those in any case.

**Comment by r00ster91** [comment]
@Detegr I made some changes and tested it. Do you mind taking a look again?

**Comment by r00ster91** [comment]
Done.

**Pull request merged**